performance/test-perf.js

```diff
@@ -0,0 +1,405 @@
+					m("a", {href: "/bar"}, "Bar")
+				)
+			),
+			m("main",
+				m("form", {onSubmit: function onSubmit() {}},
+					m("input", {type: "checkbox", checked: true}),
+					m("input", {type: "checkbox", checked: false}),
+					m("fieldset",
+						m("label",
+							m("input", {type: "radio", checked: true})
+						),
+						m("label",
+							m("input", {type: "radio"})
+						)
+					),
+					m("button-bar",
+						m("button",
+							{style: "width:10px; height:10px; border:1px solid #FFF;"},
+							"Normal CSS"
+						),
+						m("button",
+							{style: "top:0 ; right: 20"},
+							"Poor CSS"
+						),
+						m("button",
+							{style: "invalid-prop:1;padding:1px;font:12px/1.1 arial,sans-serif;", icon: true},
+							"Poorer CSS"
+						),
+						m("button",
+							{style: {margin: 0, padding: "10px", overflow: "visible"}},
+							"Object CSS"
+						)
+					)
+				)
+			)
+		);
+
+		benchmark(
+			function () {
+				m.render(scratch, vdom);
+			},
+			verify("rerender without changes", 5, done)
+		);
+	});
+
+	o("repeated trees", function (done, timeout) {
+		timeout(5000);
+
+		var Header = {
+			view : function () {
+				return m("header",
+					m("h1", {"class": "asdf"}, "a ", "b", " c ", 0, " d"),
+					m("nav",
+						m("a", {href: "/foo"}, "Foo"),
+						m("a", {href: "/bar"}, "Bar")
+					)
+				);
+			}
+		}
+
+		var Form = {
+			view : function () {
+				return m("form", {onSubmit: function onSubmit() {}},
+					m("input", {type: "checkbox", checked: true}),
+					m("input", {type: "checkbox", checked: false}),
+					m("fieldset",
+						m("label",
+							m("input", {type: "radio", checked: true})
+						),
+						m("label",
+							m("input", {type: "radio"})
+						)
+					),
+					m(ButtonBar, null)
+				);
+			}
+		}
+
+		var ButtonBar = {
+			view : function () {
+				return m("button-bar",
+					m(Button,
+						{style: "width:10px; height:10px; border:1px solid #FFF;"},
+						"Normal CSS"
+					),
+					m(Button,
+						{style: "top:0 ; right: 20"},
+						"Poor CSS"
+					),
+					m(Button,
+						{style: "invalid-prop:1;padding:1px;font:12px/1.1 arial,sans-serif;", icon: true},
+						"Poorer CSS"
+					),
+					m(Button,
+						{style: {margin: 0, padding: "10px", overflow: "visible"}},
+						"Object CSS"
+					)
+				);
+			}
+		}
+
+		var Button = {
+			view : function (vnode) {
+				return m("button", vnode.attrs, vnode.children);
+			}
+		}
+
+		var Main = {
+			view : function () {
+				return m(Form);
+			}
+		}
+
+		var Root = {
+			view : function () {
+				return m("div",
+					{"class": "foo bar", "data-foo": "bar", p: 2},
+					m(Header, null),
+					m(Main, null)
+				);
+			}
+		}
+
+		var Empty = {
+			view : function () {
+				return m("div");
+			}
+		}
+
+		var Parent = {
+			view : function (vnode) {
+				return m(vnode.attrs.child);
+			}
+		}
+
+		benchmark(
+			function () {
+				m.render(scratch, m(Parent, {child : Root}));
+				m.render(scratch, m(Parent, {child : Empty}));
+			},
```

Comment:
The nodes in the recycling pool are discarded after one redraw where they were not used:

http://jsbin.com/giqudojumo/edit?html,js,console

If you render `[m('p')])` then `[]` then `[m('p')]`, Mithril recycles the original DOM node because it was in the pool. If you do `[m('p')])` => `[]` => `[]` => `[m('p')]`, the second empty list doesn't have the original `<p>` in its pool (it has no pool actually) and Mithril creates a new DOM node.

The `Parent` dance I asked you to change was actually busting the pool (which only works in list context), so my comment was partly misguided... But it would be interesting to have it both pooled and unpooled anyway.